providers/azure.go

```diff
@@ -72,59 +81,243 @@ func getAzureHeader(accessToken string) http.Header {
+	if err == nil {
+		return email, err
+	}
+
 	email, err = json.Get("mail").String()
 
 	if err != nil || email == "" {
 		otherMails, otherMailsErr := json.Get("otherMails").Array()
 		if len(otherMails) > 0 {
 			email = otherMails[0].(string)
+			err = otherMailsErr
 		}
-		err = otherMailsErr
 	}
 
 	return email, err
 }
 
-// GetEmailAddress returns the Account email address
-func (p *AzureProvider) GetEmailAddress(s *sessions.SessionState) (string, error) {
-	var email string
+func getUserIDFromJSON(json *simplejson.Json) (string, error) {
+	// Try to get user ID
+	// if not defined, return empty string
+
+	uid, err := json.Get("id").String()
+	if err != nil {
+		return "", err
+	}
+
+	return uid, err
+}
+
+func (p *AzureProvider) GetUserDetails(s *sessions.SessionState) (*UserDetails, error) {
 	var err error
 
 	if s.AccessToken == "" {
-		return "", errors.New("missing access token")
+		return nil, errors.New("missing access token")
 	}
 	req, err := http.NewRequest("GET", p.ProfileURL.String(), nil)
 	if err != nil {
-		return "", err
+		return nil, err
 	}
 	req.Header = getAzureHeader(s.AccessToken)
 
 	json, err := requests.Request(req)
 
 	if err != nil {
-		return "", err
+		return nil, err
 	}
 
-	email, err = getEmailFromJSON(json)
-
-	if err == nil && email != "" {
-		return email, err
+	logger.Printf(" JSON: %v", json)
+	for key, value := range json.Interface().(map[string]interface{}) {
+		logger.Printf("\t %20v : %v", key, value)
 	}
+	email, err := getEmailFromJSON(json)
 
-	email, err = json.Get("userPrincipalName").String()
+	if err != nil {
+		logger.Printf("[GetUserDetails] failed making request: %s", err)
+		return nil, err
+	}
 
+	uid, err := getUserIDFromJSON(json)
 	if err != nil {
-		logger.Printf("failed making request %s", err)
-		return "", err
+		logger.Printf("[GetUserDetails] failed to get User ID: %s", err)
 	}
 
 	if email == "" {
 		logger.Printf("failed to get email address")
-		return "", err
+		return nil, errors.New("Client email not found")
 	}
+	logger.Printf("[GetUserDetails] Chosen email address: '%s'", email)
+	return &UserDetails{
+		Email: email,
+		UID:   uid,
+	}, nil
+}
 
-	return email, err
+// Get list of groups user belong to. Filter the desired names of groups (in case of huge group set)
+func (p *AzureProvider) GetGroups(s *sessions.SessionState, f string) (map[string]string, error) {
+	// Azure App Registration requires setting groupMembershipClaims to include group membership in identity token
+	// This option is available through ARM template only.
+	// For details refer to: https://docs.microsoft.com/pl-pl/azure/active-directory/develop/reference-app-manifest
+	if s.IDToken == "" {
+		return map[string]string{}, errors.New("missing id token")
+	}
+
+	type GroupClaims struct {
+		Groups []string `json:"groups"`
+		jwt.StandardClaims
+	}
+
+	claims := &GroupClaims{}
+	jwt.ParseWithClaims(s.IDToken, claims, func(token *jwt.Token) (interface{}, error) {
```

Comment:
Could you please explain what this anonymous function is and why it returns `empty`, could maybe do with a comment on this